File: javascript/medium/combinationSum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const res = new Array(), temp = new Array();
    generateCombinations(0, target, candidates, temp, res);
    return res;
};

var generateCombinations = function(index, target, candidates, temp, res) {
    if(target === 0) {
        res.push(temp.slice(0));
    } else if(target > 0) {
        for(let i = index; i < candidates.length; i++) {
            temp.push(candidates[i]);
            generateCombinations(i, target-candidates[i], candidates, temp, res);
            temp.pop(candidates[i]);
        }
    }
};

File: ruby/medium/combination_sum.rb
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum(candidates, target)
    res = Array.new
    temp = Array.new
    generate_combination_sum(0, target, candidates, temp, res)
    return res
end

def generate_combination_sum(index, target, candidates, temp, res)
    if(target == 0)
        res.push(temp.clone)
    elsif(target > 0)
        i = index
        while i < candidates.length
            temp.push(candidates[i])
            generate_combination_sum(i, (target - candidates[i]), candidates, temp, res)
            temp.pop
            i += 1
        end
    end
end
